File: paleo_workbench/ui/map_export_worker.py
"""Off-GUI unified-map PNG export (throwaway Fallback backend + OwnedWorkerJob)."""

from __future__ import annotations

import threading
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Mapping

from PySide6.QtCore import QObject, Signal
from PySide6.QtGui import QImage, QPainter

from paleo_workbench.mapping.map_render_backend import (
    FallbackMapRenderBackend,
    MapRenderSnapshot,
)
from paleo_workbench.ui.owned_worker_job import OwnedWorkerJob
from paleo_workbench.ui.unified_map_canvas import (
    UnifiedMapCanvas,
    paint_map_decorations,
)


@dataclass(frozen=True, slots=True)
class MapExportSpec:
    snapshot: MapRenderSnapshot
    extent: tuple[float, float, float, float]
    width: int
    height: int
    dpi: float
    decorations: Mapping[str, Any]
    path: str
    # #923: True when the LIVE canvas backend is the QGIS renderer, so the
    # worker must interpret the same renderer payloads instead of the legacy
    # painter vocabulary (which silently drops qgis_style.renderer_xml).
    prefer_native_renderer: bool = False


def unified_map_canvas_from(widget: object) -> UnifiedMapCanvas | None:
    """Return the UnifiedMapCanvas on ``widget`` or a common host wrapper."""
    if isinstance(widget, UnifiedMapCanvas):
        return widget
    for name in ("canvas", "unified_canvas", "widget"):
        inner = getattr(widget, name, None)
        if isinstance(inner, UnifiedMapCanvas):
            return inner
    return None


def snapshot_map_export(
    canvas: UnifiedMapCanvas,
    path: str | Path,
    *,
    width: int = 2400,
    height: int | None = None,
    dpi: float = 300.0,
) -> MapExportSpec:
    """Copy backend inputs on the GUI thread for a throwaway worker render."""
    extent = tuple(float(value) for value in canvas.view_extent)
    if len(extent) != 4:
        raise ValueError("view extent must contain four values")
    if height is None:
        xmin, ymin, xmax, ymax = extent
        if xmax > xmin and ymax > ymin:
            height = max(64, min(16000, round(int(width) * (ymax - ymin) / (xmax - xmin))))
        else:
            height = 1600
    if int(width) < 1 or int(height) < 1:
        raise ValueError("export size must be positive")
    provider = getattr(canvas, "_overlay_provider", None)
    state = provider() if provider is not None else {}
    decorations = dict((state or {}).get("decorations") or {})
    backend = getattr(canvas, "backend", None)
    return MapExportSpec(
        snapshot=canvas.backend._snapshot,
        extent=(extent[0], extent[1], extent[2], extent[3]),
        width=int(width),
        height=int(height),
        dpi=float(dpi),
        decorations=decorations,
        path=str(path),
        prefer_native_renderer=getattr(backend, "backend_name", "") == "qgis",
    )


def _render_frame_native(spec: MapExportSpec):
    """Render the snapshot through a throwaway QGIS backend on this thread.

    The bridge's ``render_sync`` blocks on ``waitForFinished()`` (no caller
    event loop needed), and every native object here is born and torn down on
    the calling worker thread, so there is no cross-thread QObject affinity.
    This keeps export on ONE style interpreter with the screen (#923): the
    legacy painter vocabulary cannot express qgis_style.renderer_xml.
    """
    from paleo_workbench.mapping.map_render_backend import QgisMapRenderBackend

    backend = QgisMapRenderBackend()
    backend.initialize()
    try:
        backend.set_layer_snapshot(spec.snapshot)
        backend.set_extent(spec.extent)
        backend.set_output_size(spec.width, spec.height)
        backend.set_dpi(spec.dpi)
        return backend.render_sync()
    finally:
        try:
            backend.shutdown()
        except Exception:  # noqa: BLE001 — teardown must not mask the export
            pass


def render_and_save_map_export(spec: MapExportSpec) -> None:
    """Render the map body with the same interpreter as the screen (#923).

    When the live canvas is the QGIS renderer the frame comes from a throwaway
    QGIS backend on this worker thread; anything that fails (bridge missing,
    scalar pipeline unready, native error) degrades to the legacy painter
    path exactly as before.
    """
    frame = None
    if spec.prefer_native_renderer:
        try:
            frame = _render_frame_native(spec)
        except Exception as exc:  # noqa: BLE001 — degrade, never lose the export
            import logging

            logging.getLogger(__name__).warning(
                "PNG 导出回退到回退渲染器（QGIS 渲染失败：%s）", exc
            )
    if frame is None:
        backend = FallbackMapRenderBackend()
        backend.initialize()
        backend.set_layer_snapshot(spec.snapshot)
        backend.set_extent(spec.extent)
        backend.set_output_size(spec.width, spec.height)
        backend.set_dpi(spec.dpi)
        frame = backend.render_sync()
    image = QImage(
        frame.rgba,
        frame.width,
        frame.height,
        frame.stride,
        QImage.Format.Format_RGBA8888,
    ).copy()
    painter = QPainter(image)
    painter.setRenderHint(QPainter.RenderHint.Antialiasing, True)
    paint_map_decorations(
        painter,
        spec.decorations,
        width=spec.width,
        height=spec.height,
        extent=spec.extent,
        dpi=spec.dpi,
    )
    painter.end()
    if not image.save(spec.path, "PNG"):
        raise RuntimeError("could not save unified map PNG")


class MapExportWorker(QObject):
    """Run ``render_and_save_map_export`` off the GUI thread."""

    finished = Signal(str)
    failed = Signal(str)
    cancelled = Signal()

    def __init__(self, spec: MapExportSpec, parent=None) -> None:
        super().__init__(parent)
        self._spec = spec
        self._cancel_event = threading.Event()

    def cancel(self) -> None:
        self._cancel_event.set()

    def run(self) -> None:
        if self._cancel_event.is_set():
            self.cancelled.emit()
            return
        try:
            render_and_save_map_export(self._spec)
        except Exception as exc:  # noqa: BLE001 — surface any failure to UI
            if self._cancel_event.is_set():
                # A render that raised mid-way may have left partial bytes
                # on disk; a cancelled export must never keep them (#852).
                self._discard_partial()
                self.cancelled.emit()
                return
            # A FAILED (non-cancelled) render may equally have written a
            # partial file before raising — never leave it behind (#937-11,
            # #852 covered cancellation only).
            self._discard_partial()
            self.failed.emit(f"{exc.__class__.__name__}: {exc}")
            return
        if self._cancel_event.is_set():
            # The render completed and wrote its file while the user was
            # cancelling: drop the stale half-product instead of reporting
            # "cancelled" next to a completed-looking file (#852).
            self._discard_partial()
            self.cancelled.emit()
            return
        self.finished.emit(self._spec.path)

    def _discard_partial(self) -> None:
        """Remove the target file written by a cancelled render (best-effort)."""
        try:
            Path(self._spec.path).unlink(missing_ok=True)
        except OSError:
            pass


def start_map_export_job(
    job: OwnedWorkerJob,
    spec: MapExportSpec,
    *,
    on_finished,
    on_failed,
    on_cancelled=None,
) -> MapExportWorker:
    worker = MapExportWorker(spec)
    connections = [
        (worker.finished, on_finished),
        (worker.failed, on_failed),
    ]
    if on_cancelled is not None:
        connections.append((worker.cancelled, on_cancelled))
    job.start(
        worker,
        terminal_signals=(worker.finished, worker.failed, worker.cancelled),
        result_connections=tuple(connections),
        cancel=worker.cancel,
        target=spec.path,
    )
    return worker

File: libs/ui_canvas/src/qt/map_export_worker.cpp
// UI-15 — map export worker (map_export_worker.py parity).

#include <pwb/ui_canvas/qt/map_export_worker.hpp>

#include <pwb/ui_canvas/qt/fallback_map_backend.hpp>

#include <QImage>
#include <QMetaType>
#include <QPainter>
#include <QWidget>

#include <pwb/job_runtime/job_scheduler.hpp>
#include <pwb/job_runtime/qt/job_bridge.hpp>
#include <pwb/ui_canvas/qt/unified_map_canvas.hpp>
#include <pwb/ui_widgets/map_chrome.hpp>
#include <pwb/ui_widgets/ui_context.hpp>

#include "json_variant.hpp"

namespace pwb::ui_canvas {

namespace {

// RenderContext.device_px_per_logical_px(dpi): physical device px per
// logical px — dpi / 96 for exports (Python parity).
double device_px_per_logical_px(double dpi) {
    return dpi > 0.0 ? dpi / 96.0 : 1.0;
}

}  // namespace

MapExportReport render_and_save_map_export(
    const MapExportSpec& spec,
    const std::function<bool()>& cancel) {
    std::optional<RenderFrame> frame;
    MapExportReport report;
    if (spec.prefer_native_renderer) {
        // _render_frame_native parity: a throwaway native backend born and
        // torn down on this worker thread — render_sync needs no caller
        // event loop. Anything that fails degrades to the fallback
        // renderer, never loses the export (#923).
        try {
            std::shared_ptr<MapRenderBackend> backend =
                create_native_map_render_backend();
            if (!backend) {
                throw std::runtime_error("qgis renderer unavailable");
            }
            try {
                backend->set_layer_snapshot(spec.snapshot);
                backend->set_extent(spec.extent);
                backend->set_output_size(spec.width, spec.height);
                backend->set_dpi(spec.dpi);
                frame = backend->render_sync();
            } catch (...) {
                try {
                    backend->shutdown();
                } catch (...) {
                }
                throw;
            }
            try {
                backend->shutdown();
            } catch (...) {
            }
            if (frame.has_value()) {
                report.engine = "qgis";  // only once a frame exists
            }
        } catch (const std::exception& exc) {
            report.degraded = true;
            report.degraded_reason = exc.what();
            frame.reset();
        } catch (...) {
            report.degraded = true;
            report.degraded_reason = "unknown native renderer error";
            frame.reset();
        }
    } else {
        report.degraded = true;
        report.degraded_reason =
            "QGIS renderer not requested for this export";
    }
    if (cancel && cancel()) {
        throw ExportCancelled();
    }
    if (!frame.has_value()) {
        // The fallback path — create_map_render_backend(prefer_qgis=false)
        // picks the first registered non-QGIS factory. Installing here is
        // what makes the fallback concrete (Python has it built into the
        // selector; static-library TUs need an explicit pull).
        qt::install_fallback_backend_factory();
        std::shared_ptr<MapRenderBackend> backend =
            create_map_render_backend(/*prefer_qgis=*/false);
        backend->set_layer_snapshot(spec.snapshot);
        backend->set_extent(spec.extent);
        backend->set_output_size(spec.width, spec.height);
        backend->set_dpi(spec.dpi);
        frame = backend->render_sync();
    }
    if (cancel && cancel()) {
        // The (non-interruptible) frame render finished while the user
        // was cancelling: stop BEFORE decorations/save instead of
        // producing a half-product (#1224 honest cancellation).
        throw ExportCancelled();
    }
    QImage image =
        QImage(frame->rgba.data(), frame->width, frame->height,
               frame->stride, QImage::Format_RGBA8888)
            .copy();
    {
        QPainter painter(&image);
        painter.setRenderHint(QPainter::Antialiasing, true);
        pwb::ui_widgets::paint_map_decorations(
            painter, qt::detail::json_to_variant_map(spec.decorations),
            spec.width, spec.height, spec.extent,
            device_px_per_logical_px(spec.dpi),
            /*dark_chrome=*/true);
    }
    if (cancel && cancel()) {
        throw ExportCancelled();
    }
    if (!image.save(QString::fromStdString(spec.path), "PNG")) {
        throw std::runtime_error("could not save unified map PNG");
    }
    return report;
}

MapExportSpec snapshot_map_export(const UnifiedMapCanvas& canvas,
                                  std::string path, int width,
                                  std::optional<int> height, double dpi) {
    MapRenderBackend* backend = canvas.backend();
    if (backend == nullptr) {
        // Python reads canvas.backend._snapshot directly — a missing
        // backend surfaces the same loud failure.
        throw std::runtime_error(
            "cannot snapshot an export without a render backend");
    }
    const Extent extent = canvas.view_extent();
    Json decorations = Json::object();
    if (canvas.overlay_provider()) {
        const Json state = canvas.overlay_provider()();
        const auto it = state.find("decorations");
        if (it != state.end() && it->is_object()) {
            decorations = *it;
        }
    }
    return make_export_spec(backend->snapshot(), extent, std::move(path),
                            width, height, dpi, std::move(decorations),
                            backend->backend_name() == "qgis");
}

UnifiedMapCanvas* unified_map_canvas_from(QWidget* widget) {
    // Duck-walk parity: the widget itself, else its first descendant
    // canvas (the Python attribute names canvas/unified_canvas/widget
    // all resolve to a descendant lookup in Qt).
    if (auto* canvas = qobject_cast<UnifiedMapCanvas*>(widget)) {
        return canvas;
    }
    return widget ? widget->findChild<UnifiedMapCanvas*>() : nullptr;
}

// ---------------------------------------------------------------------------
// MapExportWorker
// ---------------------------------------------------------------------------

MapExportWorker::MapExportWorker(MapExportSpec spec, QObject* parent)
    : QObject(parent), spec_(std::move(spec)) {
    // One background lane — Python's dedicated OwnedWorkerJob thread.
    pwb::job::JobScheduler::Options options;
    options.max_workers = 1;
    options.interactive_workers = 0;
    scheduler_ = std::make_unique<pwb::job::JobScheduler>(options);
    job_ = new pwb::job::qtbridge::JobOwner(this);
}

MapExportWorker::~MapExportWorker() { shutdown(0); }

void MapExportWorker::start() {
    if (started_) {
        return;
    }
    started_ = true;
    const MapExportSpec spec = spec_;
    pwb::job::JobSpec job;
    job.kind = "io";
    job.title = "unified map export";
    job.run = [spec](pwb::job::JobContext& ctx) -> std::any {
        // Worker.run() parity:
        //   pre-start cancel check → cancelled
        //   failure while cancelled → cancelled; otherwise → failed
        //   success + cancel raced → discard partial + cancelled
        const auto cancelled = [&ctx]() {
            return ctx.token().is_cancelled();
        };
        if (cancelled()) {
            ctx.check_cancelled();  // → JobCancelled
        }
        try {
            MapExportReport report =
                render_and_save_map_export(spec, cancelled);
            if (cancelled()) {
                // The render completed and wrote its file while the user
                // was cancelling: drop the stale half-product (#852).
                discard_partial_export_file(spec.path);
                ctx.check_cancelled();
                throw pwb::job::JobCancelled(ctx.job_id());
            }
            return report;
        } catch (const ExportCancelled&) {
            discard_partial_export_file(spec.path);
            ctx.check_cancelled();
            throw pwb::job::JobCancelled(ctx.job_id());
        } catch (...) {
            // A render that raised mid-way may have left partial bytes on
            // disk; a cancelled or failed export must never keep them
            // (#852 / #937-11).
            discard_partial_export_file(spec.path);
            if (cancelled()) {
                ctx.check_cancelled();
                throw pwb::job::JobCancelled(ctx.job_id());
            }
            throw;
        }
    };
    job_->start(*scheduler_, std::move(job),
                [this](const pwb::job::qtbridge::JobOutcome& outcome) {
                    switch (outcome.state) {
                    case pwb::job::JobState::done:
                    case pwb::job::JobState::degraded: {
                        const auto* report =
                            std::any_cast<MapExportReport>(&outcome.result);
                        if (report != nullptr) {
                            report_ = *report;
                        }
                        emit finished(
                            QString::fromStdString(spec_.path));
                        break;
                    }
                    case pwb::job::JobState::failed:
                        emit failed(
                            QString::fromStdString(outcome.error));
                        break;
                    case pwb::job::JobState::cancelled:
                        emit cancelled();
                        break;
                    default:
                        break;
                    }
                });
}

void MapExportWorker::cancel() {
    job_->cancel();  // cooperative — threading.Event.set parity
}

bool MapExportWorker::is_running() const { return job_->is_running(); }

bool MapExportWorker::shutdown(int wait_ms) {
    const bool joined = job_->shutdown(wait_ms);
    scheduler_->shutdown(false, wait_ms / 1000.0);
    return joined;
}

}  // namespace pwb::ui_canvas
